Clear only plugin cache if WC header are not present to load with WC headers.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

WC headers are added in filter `extra_plugin_headers`, however, in case when WooCommerce is activated/updated, `get_plugins` will be called and cache will be set before this filter can be cached.

Earlier we were also clearing the update cache, which was causing issue with showing update notices (see #26348), this only clears the plugin cache.

Also, `get_plugins` call is expensive even with an update cache present, because it does many I/O calls, so we should clear the cache very conservatively.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #26348.

### How to test the changes in this Pull Request:

1. Install an outdated plugin ([example](https://downloads.wordpress.org/plugin/akismet.4.1.zip)) (you can download the previous version of a plugin from WordPress.org > Plugin page > Advanced view ).
2. Update to WooCommerce 4.1 RC2 from any older WooCommerce version
3. Make sure that update prompt for your outdated plugin is still displayed.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Clear only plugin cache after check if headers are not present.

vedanshujain committed:
Clear plugin cache if WC header are not present to load with WC headers.

WC headers are added in filter `extra_plugin_headers`, however, in case when WooCommerce is activated/updated, `get_plugins` will be called and cache will be set before this filter can be cached.

Also, `get_plugins` call is expensive even with update cache present, so we should clear it very conservatively.

vedanshujain committed:
Add extra comments

vedanshujain committed:
Add unit test and a sample Woo plugin file

vedanshujain commented:
Note: We introduced the original change in #26358.

ObliviousHarmony commented on tests/php/includes/admin/helper/class-tests-wc-helper.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Class WC_Tests_WC_Helper.
+ */
+class WC_Tests_WC_Helper extends WC_Unit_Test_Case {
```
Since this is a test for legacy code in the `includes` directory, this test actually belongs in the `legacy` test directory!

ObliviousHarmony commented on tests/legacy/bootstrap.php:
```diff
@@ -44,6 +44,15 @@ public function __construct() {
+		if ( file_exists( WP_PLUGIN_DIR . '/sample-woo-plugin.php' ) ) {
+			unlink( WP_PLUGIN_DIR . '/sample-woo-plugin.php' );
+		}
+		copy( __DIR__ . '/data/sample-woo-plugin.php', WP_PLUGIN_DIR . '/sample-woo-plugin.php' );
```
How would you feel about doing this in the test's `setUp` and `tearDown` instead? This isn't something every test needs. I recognize this is likely done in a tmp directory, but it doesn't HAVE to!

vedanshujain replied:
Makes sense, I have moved this logic to test's body instead of setUp though because not all tests will use this boilerplate. Ref: https://github.com/woocommerce/woocommerce/pull/26358/commits/f4e169f21e2c4bee3c17c0160ac2b1b4cc21cff5

ObliviousHarmony commented on sample-woo-plugin.php:
```diff
@@ -0,0 +1,6 @@
+<?php // phpcs:ignoreFile
```
Accidentally added by a previous `copy` iteration?

vedanshujain replied:
Yup, removed now in https://github.com/woocommerce/woocommerce/pull/26358/commits/28fd8850fa84b42d38d116e7aed35c9d0d537322

vedanshujain committed:
Deleted accidently added file

denho commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1149,9 +1149,19 @@ public static function get_local_woo_plugins() {
+		 * WC Headers will not be present if `wc_enable_wc_plugin_headers` hook was added after a `get_plugins` call -- for example when WC is activated/updated.
+		 * Also, get_plugins call is expensive so we should clear this cache very conservatively.
+		 */
+		$first_plugin = array_keys( $plugins )[0];
```
It could be worth adding a check if `$plugins` is not empty.

vedanshujain replied:
I am hoping that `$plugins` will at least contain WooCommerce. Is there an edge case where this won't happen?

denho reviewed (approved):
Thanks, @vedanshujain. LGTM, approved!

ObliviousHarmony commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1149,9 +1149,19 @@ public static function get_local_woo_plugins() {
+		 * WC Headers will not be present if `wc_enable_wc_plugin_headers` hook was added after a `get_plugins` call -- for example when WC is activated/updated.
+		 * Also, get_plugins call is expensive so we should clear this cache very conservatively.
+		 */
+		$first_plugin = array_keys( $plugins )[0];
```
I think we can simplify this a bit since we can be sure that the `$plugins` array is never empty. What do you think of something like `if ( ! array_key_exists( 'Woo', reset( $plugins ) ) ) {`?

vedanshujain replied:
Agreed, added in https://github.com/woocommerce/woocommerce/pull/26358/commits/fd6c5f34ce9d58577319446f1ae7f47cc0df47ff.

ObliviousHarmony reviewed (changes requested):
I'm a big fan of the caching being done to minimize the directory scanning! There's just a few comments!

vedanshujain committed:
Moved test specific logic from bootstrap to test file

vedanshujain committed:
Simplified logic for better readability.

ObliviousHarmony reviewed (approved):
Noting that I'm approving the code changes, but the test file location will need examined next week!

vedanshujain committed:
Add back phpcs for legacy because we still want it tested.

There are lots of tests in legacy folder which we might update from time to time. Its still a good idea to keep phpcs for the folder.

vedanshujain committed:
Rename file and class to resemble PSR4 like convention for consistency.

vedanshujain committed:
Add exception for tests/php for filename sniff because it conflicts with PSR4.

vedanshujain commented:
@ObliviousHarmony I have made some more changes -- I have renamed file and class name to resemble PSR4. Also, since we actually are not enforcing PSR4 in tests/php folder, I have not changed the file location because searching/adding tests in the legacy folder was pretty un-intuitive. 

We can add tests for files in `src` directory in `tests/php/src` folder. Let me know what you think?

ObliviousHarmony commented:
Sorry for letting this slip @vedanshujain! I think this is a great balance. Let's also go ahead and change the PSR-4 namespace to the `tests/php/src` folder you've mentioned, and then that should be great!

vedanshujain commented:
@ObliviousHarmony Make sense. I am merging this PR then and starting another PR so that we have separation of scope and we can include this PR in 4.2 milestones.

vedanshujain commented:
@ObliviousHarmony Added new PR, which right now contains changes only for readme.md. I don't think any other changes are necessary. Let me know if you have something more in mind: https://github.com/woocommerce/woocommerce/pull/26451.

vedanshujain commented:
Also, removing change log label because it was already added in 4.1
